### PR TITLE
Use more smart pointers

### PR DESCRIPTION
**Pull request by kintel**
This introduces smart pointers and make_unique/make_shared in a lot of places where we previously create objects with `new` and passed naked pointers around.

Very minimal cleanup has been done apart from the smart pointers, but there may be opportunities now to refactor some convoluted pieces of code.

### CODE REVIEW

File: src/geometry/GeometryEvaluator.cc [hunk 1]
```diff
@@ -118,7 +115,8 @@ GeometryEvaluator::ResultObject GeometryEvaluator::applyToChildren(const Abstrac
   for (const auto& item : this->visitedchildren[node.index()]) {
     if (!isValidDim(item, dim)) break;
   }
-  if (dim == 2) return {applyToChildren2D(node, op)};
+  // FIXME: It's annoying to have to disambiguate here:
+  if (dim == 2) return {std::shared_ptr<Geometry>(applyToChildren2D(node, op))};
```
Comment (pca006132) [inline src/geometry/GeometryEvaluator.cc:119]: ideally we should remove this kind of casts, but we probably need some cursed metaprogramming stuff to do that...

File: src/geometry/boolean_utils.cc [hunk 2]
```diff
@@ -119,7 +119,7 @@ shared_ptr<const Geometry> applyMinkowski(const Geometry::Geometries& children)
   assert(children.size() >= 2);
   auto it = children.begin();
   t_tot.start();
-  shared_ptr<const Geometry> operands[2] = {it->second, shared_ptr<const Geometry>()};
+  shared_ptr<const Geometry> operands[2] = {it->second, std::shared_ptr<const Geometry>()};
```
Comment (pca006132) [inline src/geometry/boolean_utils.cc:122]: std?

File: src/geometry/cgal/cgalutils-applyops.cc [hunk 3]
```diff
@@ -149,293 +149,9 @@ shared_ptr<const Geometry> applyOperator3D(const Geometry::Geometries& children,
     std::string opstr = op == OpenSCADOperator::INTERSECTION ? "intersection" : op == OpenSCADOperator::DIFFERENCE ? "difference" : op == OpenSCADOperator::UNION ? "union" : "UNKNOWN";
     LOG(message_group::Error, "exception in CGALUtils::applyBinaryOperator %1$s: %2$s", opstr, e.what());
   }
-  return shared_ptr<Geometry>(N);
+  return N;
 }
 
-bool applyHull(const Geometry::Geometries& children, PolySet& result)
```
Comment (pca006132) [inline src/geometry/cgal/cgalutils-applyops.cc:155]: wow, is this `applyHull` never being used?

File: src/geometry/PolySet.h [hunk 4]
```diff
@@ -37,7 +37,7 @@ class PolySet : public Geometry
 
   bool is_convex() const;
   boost::tribool convexValue() const { return this->convex; }
-  void reset(const PolySet *ps);
+  void reset(const PolySet &ps);
```
Comment (pca006132) [inline src/geometry/PolySet.h:40]: maybe we can try to use the copy constructor?

File: src/geometry/cgal/cgalutils-mesh.cc [hunk 5]
```diff
@@ -63,8 +63,8 @@ bool createPolySetFromMesh(const TriangleMesh& mesh, PolySet& ps)
       builder.appendVertex(builder.vertexIndex(Vector3d(x, y, z)));
     }
   }
-  builder.append(&ps);
-  ps.reset(builder.build().release());
+  builder.append(ps);
+  ps.reset(*builder.build());
```
Comment (pca006132) [inline src/geometry/cgal/cgalutils-mesh.cc:67]: ideally we should be able to remove this reset.

File: src/io/export.cc [hunk 6]
```diff
@@ -232,16 +232,20 @@ bool ExportMesh::foreach_triangle(const std::function<bool(const std::array<std:
...
-  ps.vertices.clear();
-  ps.vertices.reserve(vertices.size());
-  ps.indices.clear();
-  ps.indices.reserve(triangles.size());
-  for (const auto& v : vertices)
-    ps.vertices.push_back({v[0], v[1], v[2]});
-  for (const auto& tri : triangles)
-    ps.indices.push_back({tri.key[0], tri.key[1], tri.key[2]});
+  auto ps = std::make_unique<PolySet>(3);
+  ps->vertices.clear();
```
Comment (pca006132) [inline src/io/export.cc:238]: no need to clear

File: src/io/import_3mf.cc [hunk 7]
```diff
@@ -262,69 +241,69 @@ Geometry *import_3mf(const std::string& filename, const Location& loc)
     }
```
Comment (pca006132) [inline src/io/import_3mf.cc:241]: forgot to modify the `return new PolySet(3);`.

File: src/io/import_3mf.cc [hunk 8]
```diff
@@ -342,33 +321,34 @@ Geometry *import_3mf(const std::string& filename, const Location& loc)
...
-    return new PolySet(3);
+    return std::make_unique<PolySet>(3);
   } else if (meshes.empty()) {
     return first_mesh;
   } else {
-    PolySet *p = nullptr;
+    std::unique_ptr<PolySet> p;
 #ifdef ENABLE_CGAL
     Geometry::Geometries children;
     children.push_back(std::make_pair(std::shared_ptr<const AbstractNode>(), shared_ptr<const Geometry>(first_mesh)));
```
Comment (pca006132) [inline src/io/import_3mf.cc:340]:
```diff
-    children.push_back(std::make_pair(std::shared_ptr<const AbstractNode>(), shared_ptr<const Geometry>(first_mesh)));
+    children.push_back(std::make_pair(
+          std::shared_ptr<const AbstractNode>(), 
+          std::shared_ptr<const Geometry>(std::unique_ptr<const Geometry>(std::move(first_mesh)))));
```
seems that the compiler need this kind of cast...

File: src/io/DxfData.h [hunk 9]
```diff
@@ -42,5 +42,5 @@ class DxfData
 
   void fixup_path_direction();
   [[nodiscard]] std::string dump() const;
-  [[nodiscard]] class Polygon2d *toPolygon2d() const;
```
Comment (pca006132) [inline src/io/DxfData.h:45]: need `#include <memory.h>`.

**Comment by pca006132** [comment]
Also, I don't think we need `std::shared_ptr<manifold::Manifold>` for `ManifoldGeometry`. manifold stores the internal data behind another shared pointer already.

File: src/core/DrawingCallback.h [hunk 10]
```diff
@@ -57,8 +57,8 @@ class DrawingCallback
   double size;
 
   Outline2d outline;
-  Polygon2d *polygon{nullptr};
-  std::vector<const Geometry *> polygons;
+  std::shared_ptr<Polygon2d> polygon;
+  std::vector<std::shared_ptr<Geometry>> polygons;
```
[the remark this answers is not in the document]
Comment (kintel) [reply src/core/DrawingCallback.h:61]: Possibly. Good reminder to go through and get `const` right.

File: src/geometry/cgal/cgalutils-applyops.cc (hunk 3 above)
Comment (kintel) [reply src/geometry/cgal/cgalutils-applyops.cc:155]: It was moved to boolean_utils earlier, and I forgot to remove it

File: src/geometry/GeometryEvaluator.cc (hunk 1 above)
Comment (kintel) [reply src/geometry/GeometryEvaluator.cc:119]: I think this is a different issue, originating our combined const/non-const ResultObject type. Ignoring this for now..

**Review by pca006132 — APPROVED** [review]
LGTM

**Pull request merged**